Updated Data Science code review (#35)
aicoe-aiops/data-science-workflows: docs/guidelines-for-reviewing-datascience-code.md

> # 10 Principles for Reviewing Data Science Code

Reviewer: Comment: I suggested some of these are too general and should be removed, so here are some Data Science specific review principles to consider instead:

* Use Code Quality tools like flake8-nb, black-nb, etc.
* Make sure credentials and  external data access is managed correctly
* Are the train/test data sets separated correctly for your ML task?
* Is the data management handled in a universal way? (no hard coding directories)    
* Is the appropriate evaluation metric being used?

Author: The suggestions are very helpful. I will edit this document incorporating all of these 👍

Reviewer 2: You can probably also expand on:

* Are the dependencies in code or environments used to run the code appropriately tracked ? eg: via Pipfile

> Code reviews are common practice in the software engineering field and have been for some time. They provide developers with a well-defined approach to sharing knowledge, ensuring exposure to new ideas and technologies, enforcing quality assurance of code and standards for the engineering process across both the codebase and the team itself. As members of the AI CoE in the office of the CTO, we believe that having a well-defined approach for how to do the same thing with data science and machine learning code is critical. Although, code is code, and we approach data science work as a subset of software engineering, there are some data science specific challenges associated with the code review process that should be addressed. Below are the 10 principles we use in our group to review data science work.

Author: That sounds food. I will be good idea. Will also add some references 👍

> Code reviews are common practice in the software engineering field and have been for some time. They provide developers with a well-defined approach to sharing knowledge, ensuring exposure to new ideas and technologies, enforcing quality assurance of code and standards for the engineering process across both the codebase and the team itself. As members of the AI CoE in the office of the CTO, we believe that having a well-defined approach for how to do the same thing with data science and machine learning code is critical. Although, code is code, and we approach data science work as a subset of software engineering, there are some data science specific challenges associated with the code review process that should be addressed. Below are the 10 principles we use in our group to review data science work.

Reviewer: @aakankshaduggal do you still plan to add the linked references?

> Code reviews are common practice in the software engineering field and have been for some time. They provide developers with a well-defined approach to sharing knowledge, ensuring exposure to new ideas and technologies, enforcing quality assurance of code and standards for the engineering process across both the codebase and the team itself. As members of the AI CoE in the office of the CTO, we believe that having a well-defined approach for how to do the same thing with data science and machine learning code is critical. Although, code is code, and we approach data science work as a subset of software engineering, there are some data science specific challenges associated with the code review process that should be addressed. Below are the 10 principles we use in our group to review data science work.

Author: Yes on it.

> 1. **Understand the PR** : When asked for a review, make sure to first understand the issue linked and the goal of the pull request. Go through the Pull Request description to understand what has changed, a checklist of prerequisites like added documentation, notebooks, etc., and any additional notes the reviewer should be aware of.

Reviewer: ```suggestion
1. **Understand the Context** : Make sure you read the pull request description and any linked issues so you know the data scientist's goal. This will help you provide appropriate level of feedback and focus on the meaningful changes. 
```

> 2. **NB Review** : Use tools like NB review when possible while reviewing Jupyter Notebooks, this will make reviewing the notebooks easier.

Reviewer: ```suggestion
2. **Rich Diffs** : Use tools like [NBreview](https://www.reviewnb.com/) or [nbdime](https://github.com/jupyter/nbdime) when possible to review notebooks. These kind of tools help you zero in on the notebook changes that actually matter. 
```

> 3. **Don’t be afraid to ask questions** : When reviewing someone’s code, if you don’t understand something or are curious about why a certain approach was implemented, ask questions, and get clarification.

Reviewer: ```suggestion
3. **Does it make sense?**: Notebooks generally contain a mix of code, markdown, plots. Do these elements reference each other correctly? If you don’t understand something or are curious about why a certain approach was implemented, ask and get clarification.
```

> 4. **Read Carefully** : Read the notebook thoroughly and try to look for logical errors or bugs in the code. Make sure you bring up any possible edge cases the code will not handle correctly.

Reviewer: ```suggestion
4. **Check the Math!** : Machine Learning code can run successfully while still being wrong and not achieving the appropriate task.  Read the notebook thoroughly and try to look for any mathematical errors in the code with reference to the PR's stated goal. 
```

> 3. **Does it make sense?** :Notebooks generally contain a mix of code, markdown, plots. Do these elements reference each other correctly? If you don’t understand something or are curious about why a certain approach was implemented, ask and get clarification.

Reviewer 3: This might be a good place to add in a comment on making sure the evaluation metric makes sense (ie, not trying to use RMSE for classification).

Author: Thanks @isabelizimm, I will add this.

> 5. **Code Optimization** : Suggest alternate approaches to perform a similar action or ways to restructure nested functions/loops if possible. Instead of having one huge notebook, always look if it is possible to modularize the notebook into smaller notebooks. This will present new techniques and various implementations for the team to consider.

Reviewer: This does not seem very specific to data science. I think this would be true about all code. Except the modular notebook bit, that said, Im not sure I agree that its _always_ best practice to divide a notebook. This section could be replaced.

Author: In my opinion, it gets hard for the reviewer to review a HUGE PR. So, maybe we could rephrase it a little or we could move this point to polishing the notebook, WDYT? @MichaelClifford

Reviewer: It is difficult for a reviewer to review a huge PR. But dividing 1 big notebook into 3 does not actually make the PR any smaller. Also, I think it is the Data Scientist's responsibility to submit reasonably sized PR's. I still think this section can probably be replaced.

Author: Sure :+1:

> 6. **Future steps and actions** : Make sure the goals of the notebook are met in the end, the key takeaways are highlighted, and should contain the data scientist’s interpretation of any statistical and/ or experimental results. While reviewing the code, it is important to ensure that future actions are mentioned so that readers or future developers picking up the work will know the developer’s intentions for follow-up experiments or implementations.

Reviewer: ```suggestion
6. **Does it solve the right problem?** : Make sure the goals of the PR or notebook are met, the key takeaways are highlighted and the data scientist’s interpretation of any statistical and/ or experimental results are included.
```

7. **Introduce new libraries or technologies** : When working on a data science project, there might be times when you are aware of a library or new technology that can address the problem in a more efficient way. In such scenarios, you should suggest this approach in the code review.

> 8. **Evaluate the output for each cell** : If you are unsure about any output or results, clone the pull request to your environment and try to run it yourself for better understanding.

Reviewer: ```suggestion
8. **Is it reproducible? **: If you are unsure about any outputs or results, clone the pull request to your environment and run it yourself. It should run from top to bottom without error and generate the same outputs as are presented in the PR
```

9. **Understand the visualizations** : When reviewing a PR that includes visualizations, make sure to check for the proper labeling of axes and ticks. All the values mentioned should be clear and have an inference for each graph.

> 10. **Readability and Code Commenting** : Well commented code makes it easier to understand what the code is doing and, more importantly, why it is doing it in a particular way. Comments are meant to be read by future programmers and should be clear and precise. If you need to pause frequently during the review yourself, it might be because readability is not that great and readability also depends on having appropriate comments in the code.

Reviewer: I think this would be true in all SE work. This could probably be removed.

> 10.  **Machine Learning Data and Evaluation** : Everything looks right, from choosing a cool preprocessing step to a brand new loss and a neat accuracy metric, but validation or test results are way worse than what was expected. Sometimes simplest things can go wrong like overfitting, or the model is wrong, or is it the loss? Make sure to double-check the smallest detail when reviewing an ML code.

Reviewer 3: A more descriptive title for this tip might be "Double Check the Details" or something similar.

Author: Yeah, I felt double check might not be very self-descriptive. WDYT? Seeking cool(er) suggestions for this one :) @isabelizimm

Reviewer 3: Maybe "Evaluate the output"?

Author: Sounds good, added that :+1:

1. **Understand the Context** : Make sure you read the pull request description and any linked issues so you know the data scientist's goal. This will help you provide an appropriate level of feedback and focus on the meaningful changes. 

2. **Rich Diffs** : Use tools like [NBreview](https://www.reviewnb.com/) or [nbdime](https://github.com/jupyter/nbdime) when possible to review notebooks. These kinds of tools help you zero in on the notebook changes that actually matter. 

> 3. **Does it make sense?** : Notebooks generally contain a mix of code, markdown, plots. Do these elements reference each other correctly? Is the evaluation metric being used here suitable for the ML algorithm ? If you don’t understand something or are curious about why a certain approach was implemented, ask and get clarification.

Reviewer 4: Small nit, and it could be just me, but "does it make sense" feels too broad. Could we maybe update this to "ensure coherence" or something along those lines?

Otherwise looks good! :smiley:

> 3. **Does it ensure coherence?** : Notebooks generally contain a mix of code, markdown, plots. Do these elements reference each other correctly? Is the evaluation metric being used here suitable for the ML algorithm ? If you don’t understand something or are curious about why a certain approach was implemented, ask and get clarification.

Reviewer: ```suggestion
3. **Is it Coherent?** : Notebooks generally contain a mix of code, markdown and plots. Do these elements reference each other correctly? Is the evaluation metric being used here suitable for the ML algorithm ? If you don’t understand something or are curious about why a certain approach was implemented, ask and get clarification.
```

4. **Check the Math!** : Machine Learning code can run successfully while still being wrong and not achieving the appropriate task. Read the notebook thoroughly and try to look for any mathematical errors in the code with reference to the PR's stated goal. 

5. **Does it solve the right problem?** : Make sure the goals of the PR or notebook are met, the key takeaways are highlighted and the data scientist’s interpretation of any statistical and/ or experimental results are included.

6. **Introduce new libraries or technologies** : When working on a data science project, there might be times when you are aware of a library or new technology that can address the problem in a more efficient way. In such scenarios, you should suggest this approach in the code review.

> 7. **Is it reproducible?** : Are the dependencies in the code or environments used to run the code appropriately tracked using a pipfile or requirements file ? If you are unsure about any outputs or results, clone the pull request to your environment and run it yourself. It should run from top to bottom without error and generate the same outputs as are presented in the PR.

Reviewer: ```suggestion
7. **Is it reproducible?** : Are all of the dependencies used to run the code appropriately tracked using a pipfile or requirements.txt ? If you are unsure about any outputs or results, clone the pull request to your environment and run it yourself. It should run from top to bottom without error and generate the same outputs as are presented in the PR.
```

8. **Understand the visualizations** : When reviewing a PR that includes visualizations, make sure to check for the proper labeling of axes and ticks. All the values mentioned should be clear and have an explanation for each graph.

9. **Hardcoded Credentials and Paths** : Sometimes developers use hardcoded credentials, secrets and even data directories for quick tests and easy access when needed and forget to remove them before publishing them to github. When reviewing a Data Science code, make sure to check for any credentials/secrets as this practice poses a significant security risk that can allow attackers to bypass authentication mechanisms and ensure that the data management is handled properly and there are no hardcoded directories in the code.

> 10.  **Evaluate the Output** : Everything looks right, from choosing a cool preprocessing step to a brand new loss and a neat accuracy metric, but validation or test results are way worse than what was expected. Sometimes simplest things can go wrong like overfitting, or the model is wrong, or is it the loss? Make sure to double-check the smallest detail when reviewing an ML code.

Reviewer: ```suggestion
10.  **Evaluate the Output** : Everything looks right, but our validation results are way worse than expected. Sometimes the simplest things can go wrong like over-fitting, poorly tuned hyper-parameters or maybe an inappropriate distance metrics being used? Make sure to double-check all the smallest detail when reviewing a ML code.
```
